programs/utilities/lockRotation.sc

```diff
@@ -0,0 +1,238 @@
+_castSide(b, raycast) ->
+(
+	blockpos = pos(b);
+	blockpos:0 += 0.5;
+	blockpos:1 += 0.5;
+	blockpos:2 += 0.5;
+	
+	difpos = [raycast:0 - blockpos:0, raycast:1 - blockpos:1, raycast:2 - blockpos:2];
+	
+	truepos = [];
+	for (difpos, truepos:_i = ceil(abs(difpos:_i) - 0.49999) * _sign(difpos:_i));
+	
+	direction = null;
+	if (truepos:0 == -1, direction = 'west');
+	if (truepos:0 ==  1, direction = 'east');
+	if (truepos:1 == -1, direction = 'down');
+	if (truepos:1 ==  1, direction = 'up');
+	if (truepos:2 == -1, direction = 'north');
+	if (truepos:2 ==  1, direction = 'south');
+	
+	return(direction);
+);
+_halfSide(half) ->
+(
+    if (half == 0, return('bottom'));
+    if (half == 1, return('top'));
+    return(null);
+);
+_contains(list, value) -> return(list ~ value != null);
+_isVertible(block) ->
+(
+    if (_contains(block, 'shulker_box'), return(true));
+    if (_contains(['end_rod', 'lightning_rod'], block), return(true));
+    return(false);
+);
+_isBustable(block) ->
+(
+    //Blocks that shouldn't be rotated lest they float or break
+    if (_contains(block, 'torch'), return(true));
+    if (_contains(block, '_bud'), return(true));
+    if (_contains(block, 'button'), return(true));
+    if (_contains(block, '_bed'), return(true));
+    if (_contains(block, 'coral_fan'), return(true));
+    if (_contains(block, 'wall_head'), return(true));
+    if (_contains(['amethyst_cluster', 'bell', 'ladder', 'lever', 'pointed_dripstone', 'tripwire_hook'], block), return(true));
+    return(false);
+);
+
+// App Methods
+_getCurrentState(outer(p)) ->
+(
+    newProfile = [null, null, null];
+    
+    direction = query(p, 'facing');
+    flat_dir = _getDir(query(p, 'yaw'));
+    
+    raycast = query(p, 'trace', 5, 'blocks', 'exact');
+	rayhit  = query(p, 'trace', 5, 'blocks');
+    if (raycast != null,
+        hit_dir = _castSide(rayhit, raycast);
+		
+        half = abs(floor((raycast:1 - floor(raycast:1)) * 2));
+        if (_dirAxis(direction) == 'y',
+            half = -1;
+        , raycast:1 < 0,
+            half = 1 - half;
+        );
+        
+		flat = hit_dir;
+		if (_contains(['up', 'down'], flat), flat = flat_dir);
+		
+        newProfile = [hit_dir, half, _flipDir(flat)];
+    ,
+        newProfile = [direction, -1, flat_dir];
+    );
+    
+    if (global_lockProfile != null && newProfile == global_lockProfile,
+        global_lockProfile = null;
+    ,
+        global_lockProfile = newProfile;
+    );
+);
+_getIdealState(block) ->
+(
+    state = block_state(block);
+    if(_isBustable(block), return(state));
+    
+    direction = global_lockProfile:0;
+    half      = global_lockProfile:1;
+    flat_dir  = global_lockProfile:2;
+    if (global_lockProfile == null, return(state));
+    
+    if (_contains(state, 'axis'), state:'axis' = _dirAxis(direction));
+    if (_contains(state, 'type'),
+        //Slab Type
+        if (_contains(['top', 'bottom'], state:'type') && half != -1,
+            state:'type' = halfSide(half);
```

Comment:
Function name should be `_halfSide` not `halfSide`

Reply:
Shoot, I thought I caught all of those. Fixed it